# Nonparametric ML Models - Cumulative Lab

## Introduction

In this cumulative lab, you will apply two nonparametric models you have just learned — k-nearest neighbors and decision trees — to the forest cover dataset.

## Objectives

* Practice identifying and applying appropriate preprocessing steps
* Perform an iterative modeling process, starting from a baseline model
* Explore multiple model algorithms, and tune their hyperparameters
* Practice choosing a final model across multiple model algorithms and evaluating its performance

## Your Task: Complete an End-to-End ML Process with Nonparametric Models on the Forest Cover Dataset

![line of pine trees](images/trees.jpg)

Photo by <a href="https://unsplash.com/@michaelbenz?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Michael Benz</a> on <a href="/s/photos/forest?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a>

### Business and Data Understanding

To repeat the previous description:

> Here we will be using an adapted version of the forest cover dataset from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/covertype). Each record represents a 30 x 30 meter cell of land within Roosevelt National Forest in northern Colorado, which has been labeled as `Cover_Type` 1 for "Cottonwood/Willow" and `Cover_Type` 0 for "Ponderosa Pine". (The original dataset contained 7 cover types but we have simplified it.)

The task is to predict the `Cover_Type` based on the available cartographic variables:

In [1]:
# Run this cell without changes
import pandas as pd

df = pd.read_csv('data/forest_cover.csv')
df

Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0           2553     235     17                               351   
1           2011     344     17                               313   
2           2022      24     13                               391   
3           2038      50     17                               408   
4           2018     341     27                               351   
...          ...     ...    ...                               ...   
38496       2396     153     20                                85   
38497       2391     152     19                                67   
38498       2386     159     17                                60   
38499       2384     170     15                                60   
38500       2383     165     13                                60   

       Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                                  95                              780   
1                                  29                              404   
2                                  42                              509   
3                                  71                              474   
4                                  34                              390   
...                               ...                              ...   
38496                              17                              108   
38497                              12                               95   
38498                               7                               90   
38499                               5                               90   
38500                               4                               67   

       Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
0                188             253            199   
1                183             211            164   
2                212             212            134   
3                226             200            102   
4                152             188            168   
...              ...             ...            ...   
38496            240             237            118   
38497            240             237            119   
38498            236             241            130   
38499            230             245            143   
38500            231             244            141   

       Horizontal_Distance_To_Fire_Points  ...  Soil_Type_31  Soil_Type_32  \
0                                    1410  ...             0             0   
1                                     300  ...             0             0   
2                                     421  ...             0             0   
3                                     283  ...             0             0   
4                                     190  ...             0             0   
...                                   ...  ...           ...           ...   
38496                                 837  ...             0             0   
38497                                 845  ...             0             0   
38498                                 854  ...             0             0   
38499                                 864  ...             0             0   
38500                                 875  ...             0             0   

       Soil_Type_33  Soil_Type_34  Soil_Type_35  Soil_Type_36  Soil_Type_37  \
0                 0             0             0             0             0   
1                 0             0             0             0             0   
2                 0             0             0             0             0   
3                 0             0             0             0             0   
4                 0             0             0             0             0   
...             ...           ...           ...           ...           ...   
38496             0             0             0             0             0   
38497             0             0             0            

> As you can see, we have over 38,000 rows, each with 52 feature columns and 1 target column:

> * `Elevation`: Elevation in meters
> * `Aspect`: Aspect in degrees azimuth
> * `Slope`: Slope in degrees
> * `Horizontal_Distance_To_Hydrology`: Horizontal dist to nearest surface water features in meters
> * `Vertical_Distance_To_Hydrology`: Vertical dist to nearest surface water features in meters
> * `Horizontal_Distance_To_Roadways`: Horizontal dist to nearest roadway in meters
> * `Hillshade_9am`: Hillshade index at 9am, summer solstice
> * `Hillshade_Noon`: Hillshade index at noon, summer solstice
> * `Hillshade_3pm`: Hillshade index at 3pm, summer solstice
> * `Horizontal_Distance_To_Fire_Points`: Horizontal dist to nearest wildfire ignition points, meters
> * `Wilderness_Area_x`: Wilderness area designation (3 columns)
> * `Soil_Type_x`: Soil Type designation (39 columns)
> * `Cover_Type`: 1 for cottonwood/willow, 0 for ponderosa pine

This is also an imbalanced dataset, since cottonwood/willow trees are relatively rare in this forest:

In [2]:
# Run this cell without changes
print("Raw Counts")
print(df["Cover_Type"].value_counts())
print()
print("Percentages")
print(df["Cover_Type"].value_counts(normalize=True))

Raw Counts
0    35754
1     2747
Name: Cover_Type, dtype: int64

Percentages
0    0.928651
1    0.071349
Name: Cover_Type, dtype: float64


Thus, a baseline model that always chose the majority class would have an accuracy of over 92%. Therefore we will want to report additional metrics at the end.

### Previous Best Model

In a previous lab, we used SMOTE to create additional synthetic data, then tuned the hyperparameters of a logistic regression model to get the following final model metrics:

* **Log loss:** 0.13031294393913376
* **Accuracy:** 0.9456679825472678
* **Precision:** 0.6659919028340081
* **Recall:** 0.47889374090247455

In this lab, you will try to beat those scores using more-complex, nonparametric models.

### Modeling

Although you may be aware of some additional model algorithms available from scikit-learn, for this lab you will be focusing on two of them: k-nearest neighbors and decision trees. Here are some reminders about these models:

#### kNN - [documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html)

This algorithm — unlike linear models or tree-based models — does not emphasize learning the relationship between the features and the target. Instead, for a given test record, it finds the most similar records in the training set and returns an average of their target values.

* **Training speed:** Fast. In theory it's just saving the training data for later, although the scikit-learn implementation has some additional logic "under the hood" to make prediction faster.
* **Prediction speed:** Very slow. The model has to look at every record in the training set to find the k closest to the new record.
* **Requires scaling:** Yes. The algorithm to find the nearest records is distance-based, so it matters that distances are all on the same scale.
* **Key hyperparameters:** `n_neighbors` (how many nearest neighbors to find; too few neighbors leads to overfitting, too many leads to underfitting), `p` and `metric` (what kind of distance to use in defining "nearest" neighbors)

#### Decision Trees - [documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html)

Similar to linear models (and unlike kNN), this algorithm emphasizes learning the relationship between the features and the target. However, unlike a linear model that tries to find linear relationships between each of the features and the target, decision trees look for ways to split the data based on features to decrease the entropy of the target in each split.

* **Training speed:** Slow. The model is considering splits based on as many as all of the available features, and it can split on the same feature multiple times. This requires exponential computational time that increases based on the number of columns as well as the number of rows.
* **Prediction speed:** Medium fast. Producing a prediction with a decision tree means applying several conditional statements, which is slower than something like logistic regression but faster than kNN.
* **Requires scaling:** No. This model is not distance-based. You also can use a `LabelEncoder` rather than `OneHotEncoder` for categorical data, since this algorithm doesn't necessarily assume that the distance between `1` and `2` is the same as the distance between `2` and `3`.
* **Key hyperparameters:** Many features relating to "pruning" the tree. By default they are set so the tree can overfit, and by setting them higher or lower (depending on the hyperparameter) you can reduce overfitting, but too much will lead to underfitting. These are: `max_depth`, `min_samples_split`, `min_samples_leaf`, `min_weight_fraction_leaf`, `max_features`, `max_leaf_nodes`, and `min_impurity_decrease`. You can also try changing the `criterion` to "entropy" or the `splitter` to "random" if you want to change the splitting logic.

### Requirements

#### 1. Prepare the Data for Modeling

#### 2. Build a Baseline kNN Model

#### 3. Build Iterative Models to Find the Best kNN Model

#### 4. Build a Baseline Decision Tree Model

#### 5. Build Iterative Models to Find the Best Decision Tree Model

#### 6. Choose and Evaluate an Overall Best Model

## 1. Prepare the Data for Modeling

The target is `Cover_Type`. In the cell below, split `df` into `X` and `y`, then perform a train-test split with `random_state=42` and `stratify=y` to create variables with the standard `X_train`, `X_test`, `y_train`, `y_test` names.

Include the relevant imports as you go.

In [3]:
# Your code here
from sklearn.model_selection import train_test_split
y = df.Cover_Type
X = df.drop('Cover_Type',axis=1)

X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=42,stratify=y)

Now, instantiate a `StandardScaler`, fit it on `X_train`, and create new variables `X_train_scaled` and `X_test_scaled` containing values transformed with the scaler.

In [4]:
# Your code here
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_train)

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

The following code checks that everything is set up correctly:

In [5]:
# Run this cell without changes

# Checking that df was separated into correct X and y
assert type(X) == pd.DataFrame and X.shape == (38501, 52)
assert type(y) == pd.Series and y.shape == (38501,)

# Checking the train-test split
assert type(X_train) == pd.DataFrame and X_train.shape == (28875, 52)
assert type(X_test) == pd.DataFrame and X_test.shape == (9626, 52)
assert type(y_train) == pd.Series and y_train.shape == (28875,)
assert type(y_test) == pd.Series and y_test.shape == (9626,)

# Checking the scaling
assert X_train_scaled.shape == X_train.shape
assert round(X_train_scaled[0][0], 3) == -0.636
assert X_test_scaled.shape == X_test.shape
assert round(X_test_scaled[0][0], 3) == -1.370

## 2. Build a Baseline kNN Model

Build a scikit-learn kNN model with default hyperparameters. Then use `cross_val_score` with `scoring="neg_log_loss"` to find the mean log loss for this model (passing in `X_train_scaled` and `y_train` to `cross_val_score`). You'll need to find the mean of the cross-validated scores, and negate the value (either put a `-` at the beginning or multiply by `-1`) so that your answer is a log loss rather than a negative log loss.

Call the resulting score `knn_baseline_log_loss`.

Your code might take a minute or more to run.

In [8]:
# Replace None with appropriate code

# Relevant imports
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score

# Creating the model
knn_baseline_model = KNeighborsClassifier()

# Perform cross-validation
knn_baseline_log_loss = -cross_val_score(knn_baseline_model,X_train_scaled,y_train,scoring="neg_log_loss").mean()

knn_baseline_log_loss

0.1255288892455634

Our best logistic regression model had a log loss of 0.13031294393913376

Is this model better? Compare it in terms of metrics and speed.

In [7]:
# Replace None with appropriate text
"""
Log loss was better with the baseline model than the tuned model. The tuned model was also much slower, the cell above took 79 seconds to run.
The business needs of the model would determine if this is an improvement or not. 
"""

'\nLog loss was better with the baseline model than the tuned model. The tuned model was also much slower, the cell above took 79 seconds to run.\nThe business needs of the model would determine if this is an improvement or not. \n'

## 3. Build Iterative Models to Find the Best kNN Model

Build and evaluate at least two more kNN models to find the best one. Explain why you are changing the hyperparameters you are changing as you go. These models will be *slow* to run, so be thinking about what you might try next as you run them.

In [9]:
from sklearn.metrics import precision_score, recall_score, accuracy_score, f1_score

In [12]:
# Your code here (add more cells as needed)
precisions = []
recalls = []
accuracies = []
f1s = []
log_losses = []

for i in list(range(1,101)):
    knn = KNeighborsClassifier(n_neighbors=i)
    knn.fit(X_train_scaled,y_train)
    pred = knn.predict(X_test_scaled)

    precisions.append((i,precision_score(y_test,pred)))
    recalls.append((i,recall_score(y_test,pred)))
    accuracies.append((i,accuracy_score(y_test,pred)))
    f1s.append((i,f1_score(y_test,pred)))

    log_loss = -cross_val_score(knn,X_train_scaled,y_train,scoring="neg_log_loss").mean()
    log_losses.append((i,log_loss))


In [15]:
k_precision = [i[1] for i in precisions]
k_recall = [i[1] for i in recalls]
k_accuracy = [i[1] for i in accuracies]
k_f1 = [i[1] for i in f1s]
k_log_loss = [i[1] for i in log_losses]

In [17]:
metrics_df = pd.DataFrame()

In [20]:
metrics_df['K'] = pd.Series(list(range(1,100)))
metrics_df['precision'] = pd.Series(k_precision)
metrics_df['recall'] = pd.Series(k_recall)
metrics_df['accuracy'] = pd.Series(k_accuracy)
metrics_df['F1 score'] = pd.Series(k_f1)
metrics_df['Log loss'] = pd.Series(k_log_loss)

In [22]:
metrics_df.describe()

K  precision     recall   accuracy   F1 score   Log loss
count  99.000000  99.000000  99.000000  99.000000  99.000000  99.000000
mean   50.000000   0.886104   0.684060   0.971238   0.770922   0.088294
std    28.722813   0.016453   0.065406   0.005003   0.045831   0.048233
min     1.000000   0.865731   0.596798   0.964991   0.708729   0.062411
25%    25.500000   0.872881   0.634643   0.967172   0.734230   0.072153
50%    50.000000   0.877953   0.662300   0.969146   0.753935   0.081310
75%    74.500000   0.899647   0.718341   0.974548   0.800975   0.090141
max    99.000000   0.946918   0.893741   0.985664   0.898975   0.495210

In [29]:
# Your code here (add more cells as needed)
metrics_df.loc[metrics_df['Log loss'] < 0.06245]

K  precision    recall  accuracy  F1 score  Log loss
17  18   0.906137  0.730713  0.975379  0.809025  0.062411

The value where K = 18, the log loss is at its least. the corresponding F1 score, accuracy precision and recall are all in the fourth quartile. So this is the value for K that I will be continuing with. 

In [30]:
eightteen_nn = KNeighborsClassifier(n_neighbors=18)
eightteen_nn.fit(X_train_scaled,y_train)

KNeighborsClassifier(n_neighbors=18)

In [34]:
metric_dict = {}
headers = {1:'Manhatten',2:"Euclidian",3:"Minkowski"}

for i in [1,2,3]:
    knn = KNeighborsClassifier(n_neighbors=18,p=i)
    knn.fit(X_train_scaled,y_train)
    pred = knn.predict(X_test_scaled)

    log_loss = -cross_val_score(knn,X_train_scaled,y_train,scoring="neg_log_loss").mean()

    metric_dict[headers[i]] = [precision_score(y_test,pred),
                         recall_score(y_test,pred),
                       accuracy_score(y_test,pred),
                             f1_score(y_test,pred),
                             log_loss]

In [37]:
p_metric_df = pd.DataFrame.from_dict(metric_dict,orient='index',columns=['Precision','Recall','Accuracy','F1 score','Log loss'])
p_metric_df

Precision    Recall  Accuracy  F1 score  Log loss
Manhatten   0.905172  0.764192  0.977457  0.828729  0.060285
Euclidian   0.906137  0.730713  0.975379  0.809025  0.062411
Minkowski   0.911275  0.717613  0.974860  0.802932  0.068926

In [38]:
p_metric_df.describe()

Precision    Recall  Accuracy  F1 score  Log loss
count   3.000000  3.000000  3.000000  3.000000  3.000000
mean    0.907528  0.737506  0.975899  0.813562  0.063874
std     0.003281  0.024021  0.001374  0.013484  0.004503
min     0.905172  0.717613  0.974860  0.802932  0.060285
25%     0.905655  0.724163  0.975119  0.805978  0.061348
50%     0.906137  0.730713  0.975379  0.809025  0.062411
75%     0.908706  0.747453  0.976418  0.818877  0.065669
max     0.911275  0.764192  0.977457  0.828729  0.068926

Manhatten distance, where p=1, shows the lowest log loss as well as the strongest recall, accuracy and F1 scores. Precision at p=1 is aproximately the mean precision of all the models. 

p=1 and k=18 are my final hyperparameter values. The log loss for these values is .0603, a significant improvement from the baseline model at 0.125

## 4. Build a Baseline Decision Tree Model

Now that you have chosen your best kNN model, start investigating decision tree models. First, build and evaluate a baseline decision tree model, using default hyperparameters (with the exception of `random_state=42` for reproducibility).

(Use cross-validated log loss, just like with the previous models.)

In [39]:
# Your code here
from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier(random_state=42)
dtc.fit(X_train,y_train)

base_log_loss = -cross_val_score(dtc,X_train,y_train,scoring="neg_log_loss").mean()
print(base_log_loss)

0.7045390124149022


Interpret this score. How does this compare to the log loss from our best logistic regression and best kNN models? Any guesses about why?

In [40]:
# Replace None with appropriate text
"""
This is a significantly higher value than the best kNN model.
Indicating the base decision tree's are not reliable compared to the kNN. 
"""

"\nThis is a significantly higher value than the best kNN model.\nIndicating the base decision tree's are not reliable compared to the kNN. \n"

## 5. Build Iterative Models to Find the Best Decision Tree Model

Build and evaluate at least two more decision tree models to find the best one. Explain why you are changing the hyperparameters you are changing as you go.

In [42]:
scaled_dtc = DecisionTreeClassifier(random_state=42)
scaled_dtc.fit(X_train_scaled,y_train)

scaled_log_loss = -cross_val_score(scaled_dtc,X_train_scaled,y_train,scoring="neg_log_loss").mean()


In [43]:
scaled_log_loss

0.7057351605151588

The log loss with scaled data was worse, not by much but still worse. So I will be tuning hyperparametes below using simply X_test. 

In [53]:
# Your code here (add more cells as needed)
criterion_log_loss = {'gini':0.0, 'entropy':0.0,"log_loss":0.0}

for c in criterion_log_loss.keys():
    clf = DecisionTreeClassifier(random_state=42,criterion=c)
    criterion_log_loss[c] = -cross_val_score(clf,X_train,y_train,scoring="neg_log_loss").mean()
    
criterion_log_loss


c:\Users\zethu\anaconda3\envs\learn-env\lib\site-packages\sklearn\model_selection\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "c:\Users\zethu\anaconda3\envs\learn-env\lib\site-packages\sklearn\model_selection\_validation.py", line 531, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\zethu\anaconda3\envs\learn-env\lib\site-packages\sklearn\tree\_classes.py", line 890, in fit
    super().fit(
  File "c:\Users\zethu\anaconda3\envs\learn-env\lib\site-packages\sklearn\tree\_classes.py", line 333, in fit
    criterion = CRITERIA_CLF[self.criterion](self.n_outputs_,
KeyError: 'log_loss'

  warnings.warn("Estimator fit failed. The score on this train-test"
c:\Users\zethu\anaconda3\envs\learn-env\lib\site-packages\sklearn\model_selection\_validation.py:548: FitFailedWarning: Estimator fit failed. The score on thi

{'gini': 0.7045390124149022, 'entropy': 0.6543002106787194, 'log_loss': nan}

Based on log loss using entropy as the criterion with non-scaled data seems to be the strongest decision tree. 

In [ ]:
# Your code here (add more cells as needed)

In [ ]:
# Your code here (add more cells as needed)

## 6. Choose and Evaluate an Overall Best Model

Which model had the best performance? What type of model was it?

Instantiate a variable `final_model` using your best model with the best hyperparameters.

In [54]:
# Replace None with appropriate code
final_model = KNeighborsClassifier(n_neighbors=18,p=1)

# Fit the model on the full training data
# (scaled or unscaled depending on the model)
final_model.fit(X_train_scaled,y_train)

KNeighborsClassifier(n_neighbors=18, p=1)

Now, evaluate the log loss, accuracy, precision, and recall. This code is mostly filled in for you, but you need to replace `None` with either `X_test` or `X_test_scaled` depending on the model you chose.

In [56]:
# Replace None with appropriate code
from sklearn.metrics import accuracy_score, precision_score, recall_score, log_loss

preds = final_model.predict(X_test_scaled)
probs = final_model.predict_proba(X_test_scaled)

print("log loss: ", log_loss(y_test, probs))
print("accuracy: ", accuracy_score(y_test, preds))
print("precision:", precision_score(y_test, preds))
print("recall:   ", recall_score(y_test, preds))

log loss:  0.06394902036640464
accuracy:  0.977456887596094
precision: 0.9051724137931034
recall:    0.7641921397379913


Interpret your model performance. How would it perform on different kinds of tasks? How much better is it than a "dummy" model that always chooses the majority class, or the logistic regression described at the start of the lab?

In [57]:
# Replace None with appropriate text
"""
The total log loss is 0.064 wich is pretty good especially considering this score for other candidate models.
The recall at 0.764 is pretty good, meaning that about 3/4 of all positive predictions made are correct. Of all 
the predictions made the model gets about 97.7% correct based on the accuracy score. The liklihood that a given 
prediction is correct is represented by precision which is also pretty good at 0.905. 

The model succesfully captures a vast majority of relevent items (tree cover type) and predicts correctly nearly
all the time. 
"""

'\nThe total log loss is 0.064 wich is pretty good especially considering this score for other candidate models.\nThe recall at 0.764 is pretty good, meaning that about 3/4 of all positive predictions made are correct. Of all \nthe predictions made the model gets about 97.7% correct based on the accuracy score. The liklihood that a given \nprediction is correct is represented by precision which is also pretty good at 0.905. \n\nThe model succesfully captures a vast majority of relevent items (tree cover type) and predicts correctly nearly\nall the time. \n'

## Conclusion

In this lab, you practiced the end-to-end machine learning process with multiple model algorithms, including tuning the hyperparameters for those different algorithms. You saw how nonparametric models can be more flexible than linear models, potentially leading to overfitting but also potentially reducing underfitting by being able to learn non-linear relationships between variables. You also likely saw how there can be a tradeoff between speed and performance, with good metrics correlating with slow speeds.